lib/Common/DataStructures/DictionaryEntry.h

```diff
@@ -167,6 +179,28 @@ namespace JsUtil
         }
     };
 
+    template <class TKey, template <class K> class THashKeyEntry>
+    class DefaultHashedKeyEntry : public THashKeyEntry<TKey>
+    {
+    public:
+        template<typename Comparer, typename TLookup>
+        inline bool KeyEquals(TLookup const& otherKey, hash_t otherHashCode)
+        {
+            return Comparer::Equals(this->Key(), otherKey);
+        }
+
+        template<typename Comparer>
+        inline hash_t GetHashCode()
+        {
+            return ((Comparer::GetHashCode(this->Key()) & 0x7fffffff) << 1) | 1;
```

Comment:
There's also a TAGHASH macro for this shift-and-set operation; I don't see any particular pattern for when we do or don't use it.